typing/btype.mli

```diff
@@ -165,11 +168,19 @@ end
 val lowest_level: int
         (* Marked type: ty.level < lowest_level *)
 val pivot_level: int
+val mirror_level: int -> int
         (* Type marking: ty.level <- pivot_level - ty.level *)
+        (* mirror_level is used for unmarking; it must be involutive,
+           invert order *)
 val mark_type: type_expr -> unit
         (* Mark a type *)
-val mark_type_node: type_expr -> unit
-        (* Mark a type node (but not its sons) *)
+val mark_type_node:
```

Comment:
I'm not a fan of this interface. Why not just use the much simpler:
```ocaml
val mark_type_node : type_expr -> bool
```
that returns `false` if the node was already marked and returns `true` and marks the node if was not already marked.

Reply:
Good idea. We should try this.

Reply:
I tried on a few cases.
While it is an improvement on the original style, it is still more verbose than the functional version.
Typically, something like
```ocaml
mark_type_node ty ~after:(it.it_do_type_expr it)
```
must be rewritten as
```ocaml
let ty = repr ty in
if mark_type_node ty then it.it_do_type_expr it ty
```
which lacks focus.
Since the marking function has to manipulate the low level representation of types, our idea was to make the code more regular, and I think the `mark_type_node ty ~after:` is successful for that. @gasche 's remark is actually a testament to that: it shows that the code is so regular that the version without `after` was only used once.

Reply:
I think that the second version you provide is noticeably clearer than the first. It is first-order and the code can mostly be understood without needing to know the definition of `mark_type_node`.

Reply:
I have finished translating the code to the `if mark_type_node ty then ...` style.
This amounts mainly to going back to the status quo ante.
The code is in the branch https://github.com/COCTI/ocaml/tree/private_type_expr_1st_order .

The main drawback is that one has to be careful with guards, as they must be executed between the check on the level and the change of the level. So in that case the abstraction provided by you `mark_type_node` does not work.
I actually ended up introducing several bugs by forgetting to do the check.
One could of course provide `markable_type_node` function, but this would just hide one comparison...

If you prefer the code in that branch, I'm ready to go with that.

Reply:
The cases without guards are much clearer. If you don't mind one last pass over the code, I think that using an interface like:
```ocaml
val is_marked_type_node : type_expr -> bool
val mark_type_node : type_expr -> unit
val try_mark_type_node : type_expr -> bool
```
where `try_mark_type_node` is equivalent to `mark_type_node` in that branch, would allow you to make the cases with guards a little clearer. I would be happy to approve the version with those changes.

Reply:
(or at least I would be happy with the parts around `mark_type_node`. I would still like to see some improvement to the code using `Internal`)

Reply:
For `Internal`, we are working on a branch `private_type_expr_setters`.
Will merge into this PR soon.

Reply:
> The cases without guards are much clearer. If you don't mind one last pass over the code, I think that using an interface like:
> 
> ```ocaml
> val is_marked_type_node : type_expr -> bool
> val mark_type_node : type_expr -> unit
> val try_mark_type_node : type_expr -> bool
> ```
> 
> where `try_mark_type_node` is equivalent to `mark_type_node` in that branch, would allow you to make the cases with guards a little clearer. I would be happy to approve the version with those changes.

Pushed a new commit to introduce the above interface, with some name changes:
```ocaml
val not_marked_node : type_expr -> bool
val mark_node : type_expr -> unit
val try_mark_node : type_expr -> bool
```